### PR TITLE
[WiP] Add Rudimentary TIN-like Behavior to HAG Filter

### PR DESCRIPTION
**Pull request by jamesmcclain**
Connects https://github.com/PDAL/PDAL/issues/2468

I am new to this community, so I do not know what the expectations are in terms of speed, robustness, and coding style.  For that reason, this PR takes a straight-forward, no-additional-dependencies approach.  The branch is built on top of a rebased version of @gadomski 's [nearest neighbor branch](https://github.com/gadomski/PDAL/tree/hag-improved).

Rather than explicitly constructing a TIN, I have taken the approach of locally calculating the Delaunay fan near each point, which should be equivalent (albeit with a parameter, the neighborhood size).  I assume that the projection (x, y, z) -> (x, y) is one-to-one and use that assumption to project then triangulate neighborhoods of ground points.  Using that assumption a second time (or at least consistent with it), the height of each point is calculated by projecting it along z onto the surface.

I am eager to receive any comments that may be forthcoming.

I marked this as WiP; I do not know all that might be missing, but I imagine that tests are, at minimum.

Input:
```json
[
    "autzen.laz",
    {
        "type":"filters.decimation",
        "step": 20
    },
    {
        "type":"filters.hag",
	"count": 30,
	"delaunay_fans": true
    },
    {
      "type":"writers.gdal",
      "filename":"autzen-hag-tin.tif",
      "dimension":"HeightAboveGround",
      "data_type":"double",
      "output_type":"mean",
      "resolution":10.0
    }
]
```

Output:

Screenshot from 2019-05-15 21-15-04

### CODE REVIEW

**Comment by chambbj** [comment]
@jamesmcclain Thanks for the contribution. I have yet to test this locally, but the results look very promising.

It would probably be good to have @plimkilde and @abellgithub to review the delaunay changes, as I'm less familiar with that portion of the codebase.

File: filters/HAGFilter.cpp [hunk 1]
```diff
@@ -56,6 +59,17 @@ std::string HAGFilter::getName() const
...
 }
 
+void HAGFilter::addArgs(ProgramArgs& args)
+{
+    args.add("count", "The number of points to fetch to determine the ground point [default: 1].",
+            m_count, point_count_t(1));
+    args.add("max_distance", "The maximum distance to the farthest nearest neighbor before the height above ground is not calculated [default: 0 (disabled)]", m_max_distance);
+    args.add("allow_extrapolation", "If true and count > 1, allow extrapolation [default: true].",
+            m_allow_extrapolation, true);
+    args.add("delaunay_fans", "Construct local Delaunay fans and infer heights from them [default: false].",
```
Comment (chambbj) [inline filters/HAGFilter.cpp]: Very minor detail, but I wonder if just `delaunay` may be easier to remember (and less to type) than `delaunay_fans`.

File: filters/private/delaunator.hpp [hunk 2]
```diff
@@ -261,15 +238,15 @@ Delaunator::Delaunator(std::vector<double> const& in_coords)
     ids.reserve(n);
 
     for (std::size_t i = 0; i < n; i++) {
-        const double x = coords[2 * i];
-        const double y = coords[2 * i + 1];
+      const double x = coords[2 * i];
```
Comment (chambbj) [inline filters/private/delaunator.hpp]: I wouldn't be opposed to a separate PR that cleans up any inconsistent code formatting, but the whitespace change in this PR detracts from the substantive changes you've made. It may be a little more clear, if you were to back these out.

**Comment by plimkilde** [comment]
The `delaunator.hpp` header is currently, as far as practically possible, copied verbatim from [delaunator-cpp](https://github.com/delfrrr/delaunator-cpp). This way, it should be simple to merge fixes and optimizations from upstream. I have a feeling that this PR is effectively creating a fork of delaunator-cpp (which is fine) but at least that is a decision that should be made consciously.

If the changes to `delaunator.hpp` are generally useful, I think you should consider submitting them as a PR to the delaunator-cpp repo. If they are too specialized for that, maybe we should keep two versions of the header around (one for the Delaunay filter, one for HAG)?

Overall this looks like an exciting contribution, I'm just worried a bit about introducing less battle-tested code to the Delaunay computation. PDAL's unit testing for the Delaunay filter is currently fairly minimal, the assumption being that stress tests for the computation have been made elsewhere. For a modified `delaunator`, we would probably need more extensive test coverage in order to avoid regressions.

File: filters/HAGFilter.cpp [hunk 3]
```diff
@@ -103,9 +137,89 @@ void HAGFilter::filter(PointView& view)
...
+            z1 = gView->getFieldAs<double>(Dimension::Id::Z, ids[0]);
+        } else if (m_delaunay_fans == false) { // Nearest-Neighbor-based interpolation
+            auto min_x = std::numeric_limits<double>::max();
+            auto max_x = std::numeric_limits<double>::min();
+            auto min_y = std::numeric_limits<double>::max();
+            auto max_y = std::numeric_limits<double>::min();
+            double weights = 0;
+            double z_accumulator = 0;
+            bool exact_match = false;
+            for (unsigned long j = 0; j < ids.size(); ++j) {
```
Comment (plimkilde) [inline filters/HAGFilter.cpp]: Why not use `size_t` for this and the other iterator variables?

File: filters/private/delaunator.hpp (hunk 2 above)
Comment (jamesmcclain) [reply filters/private/delaunator.hpp]: Certainly, will address

File: filters/HAGFilter.cpp (hunk 1 above)
Comment (jamesmcclain) [reply filters/HAGFilter.cpp]: Okay, will change

**Comment by jamesmcclain** [comment]
> The delaunator.hpp header is currently, as far as practically possible, copied verbatim from delaunator-cpp. This way, it should be simple to merge fixes and optimizations from upstream. I have a feeling that this PR is effectively creating a fork of delaunator-cpp (which is fine, but at least that is a decision that should be made consciously).

Yes, that is a fine point

> If the changes to delaunator.hpp are generally useful, I think you should consider submitting them as a PR to the delaunator-cpp repo. If they are too specialized for that, maybe we should keep two versions of the header around (one for the Delaunay filter, one for HAG)?

The changes here were just motivated by some linking problems that I encountered.  The non-inlined
code within the header lead to "multiple definition" errors when the header was included in a second place.  If there is a cleaner way to solve this, I am more than happy to use it.

> Overall this looks like an exciting contribution, I'm just worried a bit about introducing less battle-tested code to the Delaunay computation. PDAL's unit testing for the Delaunay filter is currently fairly minimal, the assumption being that stress tests for the computation have been made elsewhere. For a modified delaunator, we would probably need more extensive test coverage in order to avoid regressions.

Yes, that makes sense.  My motivation for the changes (which amount to adding the keyword `inline` in a few places and moving the body of the constructor) was pure expedience -- they are not conceptually essential at all.

File: filters/HAGFilter.cpp (hunk 3 above)
Comment (jamesmcclain) [reply filters/HAGFilter.cpp]: Certainly, will change.

File: filters/private/delaunator.hpp (hunk 2 above)
Comment (jamesmcclain) [reply filters/private/delaunator.hpp]: Apropos of that, I have some general questions about the mechanics of the project (e.g. is there a mechanical way to achieve the desired style [like some set of flags to some pretty printer]?).  Where is the best place to ask those types of questions?

Comment (chambbj) [reply filters/private/delaunator.hpp]: Some of us will use either astyle or clang-format (I prefer the latter), both of which have style files in the root directory, but there is no standard. As much as anything, we generally like to keep a file internally consistent. Running either tool on this file _might_ introduce a ton of formatting changes. When that's done, it's typically performed separately from any functional changes for clarity.

The mailing list or Gitter would be reasonable places to ask such questions.

Comment (jamesmcclain) [reply filters/private/delaunator.hpp]: Okay, many thanks

File: filters/HAGFilter.cpp (hunk 1 above)
Comment (jamesmcclain) [reply filters/HAGFilter.cpp]: Okay, addressed

File: filters/private/delaunator.hpp (hunk 2 above)
Comment (jamesmcclain) [reply filters/private/delaunator.hpp]: Okay, I have reverted all changes to `delaunator.hpp` and instead re-declared the public interface in `filters/private/delaunator-decl.hpp`.  This is somewhat unattractive, but seems like it might be the best option for now?

File: filters/HAGFilter.cpp (hunk 3 above)
Comment (jamesmcclain) [reply filters/HAGFilter.cpp]: Okay, have changed

**Comment by jamesmcclain** [comment]
I think that all review comments (at time of writing) have been addressed.  The most notable change is that I have reverted all chanages to `filters/private/delaunator.hpp` and re-declared the public interface in `filters/private/delaunator-decl.hpp`.

**Comment by abellgithub** [comment]
I'll take a look at this on Monday.  I really appreciate the work.

**Comment by plimkilde** [comment]
> The changes here were just motivated by some linking problems that I encountered. The non-inlined
> code within the header lead to "multiple definition" errors when the header was included in a second place. If there is a cleaner way to solve this, I am more than happy to use it.
> 
Your linking problems sound like a compelling reason to split the upstream delaunator-cpp into separate header and source files. I see Andrew already created [an issue](https://github.com/delfrrr/delaunator-cpp/issues/17) for that.

File: filters/HAGFilter.cpp [hunk 4]
```diff
@@ -99,13 +139,94 @@ void HAGFilter::filter(PointView& view)
...
+                double bx = gView->getFieldAs<double>(Dimension::Id::X, ids[bi]);
+                double by = gView->getFieldAs<double>(Dimension::Id::Y, ids[bi]);
+                double bz = gView->getFieldAs<double>(Dimension::Id::Z, ids[bi]);
+                double cx = gView->getFieldAs<double>(Dimension::Id::X, ids[ci]);
+                double cy = gView->getFieldAs<double>(Dimension::Id::Y, ids[ci]);
+                double cz = gView->getFieldAs<double>(Dimension::Id::Z, ids[ci]);
+
+                z1 = std::min(z1, distance_along_z(ax, ay, az, bx, by, bz, cx, cy, cz, x0, y0));
+            }
+            if (z1 == std::numeric_limits<double>::infinity())
```
Comment (abellgithub) [inline filters/HAGFilter.cpp]: I don't understand this.  It seems we're initializing z1 with infinity.  Then I would think the minimum of infinity and anything else would be anything else.  If I'm not misunderstanding, simpler logic would just be to assign to z1 as:

```
std::min(
    gView->getFieldAs<double>(Dimension::Id::Z, ids[0]),
    distance_along_z(ax, ay, az, bx, by, bz, cx, cy, cz, x0, y0)
);
